# **ScreenGenius:** Smart Movie Suggestions


![front_image](images/ipnb_front.jpg)

## Table of Contents

* [Introduction](#Introduction)

    * [Project Significance](#project-significance)
    * [The MovieLens Dataset](#the-movielens-dataset)
    * [Project Objective](#project-objective)
    * [Key Challenges](#key-challenges)
    * [Expected Outcomes](#expected-outcomes)
    * [Project Scope](#project-scope)
    * [Audience](#audience)
    * [Structure of the Report](#structure-of-the-report)

* [Problem Statement](#problem-statement)

    * [Specific Objectives](#specific-objectives)

* [Business Understanding](#business-understanding)
    * [Success Metrics](#success-metrics)

* [Data Understanding](#data-understanding)
    * [Overview](#overview)
    * [Dataset Details](#dataset-details)
    * [Data Quality and Completeness](#data-quality-and-completeness)
    * [Data Usage Considerations](#data-usage-considerations)
    * [Next Steps](#next-steps)





## Introduction

In an era where digital entertainment platforms are booming, the ability to provide users with personalized content recommendations has become a defining factor in user satisfaction and platform success. This project delves into the world of movie recommendation systems, leveraging the rich "ml-latest-small" dataset from MovieLens.

### Project Significance

Personalized movie recommendations can significantly enhance user engagement and retention on streaming platforms. By tailoring movie suggestions to individual user preferences, platforms can keep users entertained, extend their stay, and ultimately thrive in a competitive market.

### The MovieLens Dataset

The cornerstone of our project is the "ml-latest-small" dataset, a treasure trove of movie ratings and information. This dataset encompasses over 100,000 user ratings, spanning 9,742 unique movies, and originating from 610 distinct users.

### Project Objective

Our mission is clear: to build a movie recommendation system that excels at providing personalized movie suggestions based on users' historical ratings. This project aims to create a system that enhances the user experience and contributes to the success of entertainment platforms.

### Key Challenges

As I embark on this journey, I anticipate challenges such as dealing with sparse data, addressing the cold start problem for new users, and optimizing recommendation accuracy. Our approach will involve collaborative filtering techniques and, potentially, a hybrid approach to overcome these obstacles.

### Expected Outcomes

We anticipate that our recommendation system will lead to improved user engagement, higher user retention rates, and a competitive edge in the market by delivering accurate and relevant movie suggestions.

### Project Scope

This project will focus on modeling and evaluating the recommendation system using the provided dataset. The deployment aspect may be explored as future work.

### Audience

This project is intended for data scientists, developers, business stakeholders, and anyone interested in the realms of recommendation systems and movie data analysis.

### Structure of the Report

In the following sections, we will delve into data understanding, preprocessing, model development, evaluation metrics, and user interface design, culminating in the creation of a robust movie recommendation system.

## Problem Statement

The project aims to develop a movie recommendation system using the MovieLens dataset, which includes user ratings and movie information. The primary objective is to create a recommendation system that provides personalized movie suggestions to users based on their historical ratings of other movies. The dataset encompasses 100,836 user ratings, 9,742 movies, and 610 unique users.

### Specific Objectives

Specifically, the project will address the following key components:

- **User-Centric Recommendations:** TO design and implement a recommendation system that tailors movie suggestions to individual users. This system will take into account the historical ratings provided by users to identify movies that align with their preferences.

- **Collaborative Filtering:** To utilize collaborative filtering techniques, such as user-user and item-item collaborative filtering, to enhance recommendation accuracy. By leveraging user interactions and similarities between movies, the system will generate precise and personalized movie recommendations.

- **Dataset Understanding:** To analyze the provided dataset, consisting of `movieId`, `title` `genres`, `userId`, `rating` and `timestamp` columns. Understand the relationships between users, movies, and their associated genres.

- **Evaluation Metrics:** To employ suitable evaluation metrics, such as Root Mean Square Error (RMSE) or Mean Absolute Error (MAE), to assess the accuracy and effectiveness of the recommendation system. Evaluate how well the system predicts user preferences and provides relevant movie suggestions.

- **Cold Start Problem:** To consider implementing a hybrid approach, combining collaborative filtering with content-based filtering, to address the "cold start" problem. This ensures that the system can recommend movies even for new users or less-rated movies.

- **User Interface:** To develop a user-friendly interface that allows users to input their movie ratings and receive personalized recommendations. Ensure seamless integration between user ratings and the recommendation engine.

## Business Understanding

This movie recommendation system is designed to benefit the following stakeholders:

- **Users**: Movie enthusiasts seeking personalized movie suggestions based on their preferences and viewing history.

- **Streaming Platform**: The platform hosting the movie recommendation system, aiming to enhance user satisfaction and engagement.

- **Content Providers**: Movie studios and content creators interested in understanding user preferences and trends to optimize their content offerings.

### Success Metrics

To measure the success and impact of our recommendation system, we will monitor several key performance indicators (KPIs), including:

- **User Engagement**: Increased user interactions, such as movie ratings, recommendations viewed, and movies watched.

- **User Retention**: A decrease in user churn rates and an increase in recurring users.

- **Recommendation Accuracy**: Improved accuracy in predicting user preferences, as measured by evaluation metrics like RMSE and MAE.

- **Content Utilization**: A rise in the utilization of less-watched or newly added content.

- **Platform Growth**: Attraction of new users and overall platform growth.


The movie recommendation system project aligns with the evolving landscape of entertainment consumption by enhancing the user experience and delivering valuable insights for decision-makers. By leveraging data and advanced algorithms, I aim to create a win-win scenario, benefiting both our users and the platform itself while staying competitive in the market.

## Data Understanding

### Overview

The **ml-latest-small** dataset is a collection of data from [MovieLens](http://movielens.org/), a movie recommendation service. It comprises user ratings and free-text tagging activities, offering insights into user preferences. This dataset contains a total of 100,836 movie ratings and 3,683 tag applications across 9,742 movies.

### Dataset Details

- **Ratings**: There are 100,836 user ratings, primarily based on a 5-star scale.
- **Tags**: The dataset includes 3,683 instances of user-generated tags for movies.
- **Users**: These data originate from 610 users who interacted with the MovieLens platform.
- **Timeframe**: The data spans a period from March 29, 1996, to September 24, 2018.
- **No Demographic Data**: Notably, this dataset does not include any demographic information about the users. Each user is identified solely by a unique numerical identifier.
- **File Structure**: The dataset is organized into four main files: **links.csv**, **movies.csv**, **ratings.csv**, and **tags.csv**.
- Out of the four datasets, only **movie.csv** and **ratings.csv**  will be used for the analysis

The dataset comprises four files of which we will use two main files:

1. **movies.csv**: This file contains information about movies and includes the following columns:
   - `movieId`: A unique numerical identifier for each movie.
   - `title`: The title of the movie.
   - `genres`: A list of genres associated with the movie, separated by the pipe (|) character.

2. **ratings.csv**: This file contains user ratings for movies and includes the following columns:
   - `userId`: A unique numerical identifier for each user.
   - `movieId`: The corresponding movie identifier, linking each rating to a specific movie.
   - `rating`: The user's numerical rating for the movie.
   - `timestamp`: A Unix timestamp indicating when the rating was provided.


### Data Quality and Completeness

- The dataset appears to be relatively clean, with no mention of missing values or data quality issues in the provided information.

- However, further exploratory data analysis (EDA) may be necessary to identify any potential data anomalies or patterns.

### Data Usage Considerations

- The dataset is intended for development purposes and may not be suitable for sharing research results or long-term studies.

- It offers a valuable resource for creating a movie recommendation system, understanding user preferences, and enhancing user engagement on a movie platform.

### Next Steps

With a clear understanding of the dataset's structure and content, the next steps involve data preprocessing, feature engineering, and the development of the movie recommendation system. This system will leverage the rich data available to provide personalized movie suggestions to users based on their historical ratings.

## Import necessary Libraries

In [1]:
# Data manipulation and analysis
import pandas as pd                 # For working with dataframes
import numpy as np                  # For numerical operations

# Data visualization
import matplotlib.pyplot as plt     # For creating plots and visualizations
import seaborn as sns               # For advanced data visualizations

# Machine learning and recommendation
from sklearn.model_selection import train_test_split  # For splitting data into training and testing sets
from sklearn import model_selection as mod
from sklearn.metrics import mean_squared_error        # For evaluating model performance
from sklearn.metrics.pairwise import cosine_similarity # For calculating similarity between items


# Collaborative filtering algorithms




# Suppress warnings (optional)
import warnings
warnings.filterwarnings('ignore')

## Load Dataset

### Load `movies.csv`

In [2]:
# Define the file path to the movies.csv file
file_path = 'dataset/movies.csv'

# Load the dataset into a Pandas DataFrame
movies_df = pd.read_csv(file_path)

# Display the first few rows of the dataset to understand its structure
movies_df.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [3]:
# Display the shape of the dataset (number of rows and columns)
print("Dataset Shape:")
print("=================")
print("Number of Rows:", movies_df.shape[0])
print("Number of Columns:", movies_df.shape[1])

Dataset Shape:
Number of Rows: 9742
Number of Columns: 3


In [4]:
# Display the data types of columns
print("Data Types of Columns:")
print("============================")
print(movies_df.dtypes)

Data Types of Columns:
movieId     int64
title      object
genres     object
dtype: object


### Load `ratings.csv`

In [5]:
# Define the file path to the movies.csv file
file_path = 'dataset/ratings.csv'

# Load the dataset into a Pandas DataFrame
ratings_df = pd.read_csv(file_path)

# Display the first few rows of the dataset to understand its structure
ratings_df.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [6]:
# Display the shape of the dataset (number of rows and columns)
print("Dataset Shape:")
print("==========================")
print("Number of Rows:", ratings_df.shape[0])
print("Number of Columns:", ratings_df.shape[1])

Dataset Shape:
Number of Rows: 100836
Number of Columns: 4


In [7]:
# Display the data types of columns
print("Data Types of Columns:")
print("============================")
print(ratings_df.dtypes)

Data Types of Columns:
userId         int64
movieId        int64
rating       float64
timestamp      int64
dtype: object


## Data Preparation

We won't use the `timestamp`, let's drop it;

### Dropping Irrelevant Columns

In [8]:
# Drop the 'timestamp' column from the ratings dataset
ratings_df = ratings_df.drop(columns=['timestamp'])

### Merge `movies_df` & `ratings_df`

In [9]:
# Merge the datasets on the 'movieId' column
merged_df = pd.merge(ratings_df, movies_df, on='movieId')

In [10]:
# Display the first few rows of the merged dataset to understand its structure
merged_df.head()

userId  movieId  rating             title  \
0       1        1     4.0  Toy Story (1995)   
1       5        1     4.0  Toy Story (1995)   
2       7        1     4.5  Toy Story (1995)   
3      15        1     2.5  Toy Story (1995)   
4      17        1     4.5  Toy Story (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1  Adventure|Animation|Children|Comedy|Fantasy  
2  Adventure|Animation|Children|Comedy|Fantasy  
3  Adventure|Animation|Children|Comedy|Fantasy  
4  Adventure|Animation|Children|Comedy|Fantasy

This project will use Collaborative Filtering (CF) as the primary recommendation technique for the modelling of movie recommendation system due to the following compelling reasons:

- **Data Availability and Relevance**

- Our dataset contains user ratings, which are a valuable source of user-item interaction data. Collaborative filtering relies on this type of data to establish patterns of user preferences. Since we have rich user rating information, CF is a natural choice.

- **Effectiveness in Identifying User Preferences**

- Collaborative filtering techniques excel at identifying similar users and items based on historical user behavior. By analyzing user ratings, CF can accurately capture user preferences and provide recommendations that align with their interests. This approach is well-suited for our goal of delivering personalized movie suggestions.

- **Scalability and Applicability**

- Collaborative filtering methods are known for their scalability. They can efficiently handle large datasets with numerous users and items, making them suitable for our movie dataset, which includes a substantial number of user ratings.

- **Ease of Implementation**

- CF techniques are relatively straightforward to implement and experiment with. This ease of implementation allows us to quickly prototype and iterate on recommendation algorithms, making the development process efficient and adaptable to project requirements.

Collaborative Filtering is an ideal technique for our movie recommendation system due to its compatibility with our data, effectiveness in capturing user preferences, scalability, and ease of implementation. We will build upon this foundation to create a recommendation system that provides users with personalized and relevant movie suggestions.

Since `genre` information, which describes the movie's genre, is typically not considered in traditional CF.By dropping the `genre` column, we simplify the dataset to contain only the essential user-item interaction data (i.e.,`userId` , `movieId` and `rating`), which aligns with the core principles of CF.

In [11]:
# Drop the "genre" column from the dataset
data = merged_df.drop(columns=['genres'])

In [12]:
# Print the first five rows of the final dataset
data.head()

userId  movieId  rating             title
0       1        1     4.0  Toy Story (1995)
1       5        1     4.0  Toy Story (1995)
2       7        1     4.5  Toy Story (1995)
3      15        1     2.5  Toy Story (1995)
4      17        1     4.5  Toy Story (1995)

In [14]:
# Display the shape of the final dataset (number of rows and columns)
print("Dataset Shape:")
print("============================")
print("Number of Rows:", data.shape[0])
print("Number of Columns:", data.shape[1])

Dataset Shape:
Number of Rows: 100836
Number of Columns: 4


In [20]:
def check_for_duplicates(data):
    """
    Check for duplicates in a dataset and provide information about them.

    Parameters:
    - data: Pandas DataFrame or NumPy array

    Returns:
    - duplicate_info: Dictionary containing information about duplicates
    """
    import pandas as pd

    if isinstance(data, pd.DataFrame):
        duplicate_info = data[data.duplicated(keep=False)]
    elif isinstance(data, pd.ndarray):
        duplicate_info = pd.DataFrame(data).duplicated(keep=False)
    else:
        return "Unsupported data type. Please provide a Pandas DataFrame or a NumPy array."

    num_duplicates = len(duplicate_info)
    if num_duplicates == 0:
        result = "No duplicates found in the dataset."
    else:
        result = f"{num_duplicates} duplicates found in the dataset."

    duplicate_info.reset_index(drop=True, inplace=True)
    return {"result": result, "duplicates": duplicate_info}

# Call the function and pass the parameter data 
duplicate_result = check_for_duplicates(data)

print("Duplicate Check Result:")
print("==================================")
print(duplicate_result["result"])

if not duplicate_result["duplicates"].empty:
    print("\nDuplicate Rows:")
    print("======================================")
    print(duplicate_result["duplicates"])

Duplicate Check Result:
No duplicates found in the dataset.


In [24]:
def check_for_missing_values(data):
    """
    Check for missing values in a dataset.

    Parameters:
    - data: Pandas DataFrame

    Returns:
    - missing_info: Dictionary containing information about missing values
    """
    import pandas as pd

    if not isinstance(data, pd.DataFrame):
        return "Unsupported data type. Please provide a Pandas DataFrame."

    missing_counts = data.isna().sum()
    total_missing = missing_counts.sum()

    if total_missing == 0:
        result = "No missing values found in the dataset."
    else:
        result = f"{total_missing} missing values found in the dataset."

    missing_info = {
        "result": result,
        "missing_counts": missing_counts,
    }

    return missing_info

# Example usage:
# Assuming 'data' is your dataset
missing_result = check_for_missing_values(data)

print("Missing Values Check Result:")
print("======================================")
print(missing_result["result"])

if missing_result["missing_counts"].sum() > 0:
    print("\nMissing Value Counts:")
    print("======================")
    print(missing_result["missing_counts"])

Missing Values Check Result:
No missing values found in the dataset.


## Modelling

### Baseline Model-Average Rating Per Movie

In [25]:
# Calculate the average rating for each movie
average_ratings = data.groupby('movieId')['rating'].mean().reset_index()

# Rename the columns for clarity
average_ratings.columns = ['movieId', 'avg_rating']

# Display the first few rows of the average_ratings DataFrame
print("Average Ratings per Movie:")
print("===========================")
print(average_ratings.head())

# Recommendation using Baseline Model

# Merge the average_ratings with the original dataset to get recommendations
recommendations = data.merge(average_ratings, on='movieId', how='left')

# Sort the recommendations by user and average rating in descending order
recommendations.sort_values(by=['userId', 'avg_rating'], ascending=[True, False], inplace=True)

# Keep only the top-rated movie as the recommendation for each user
top_recommendations = recommendations.groupby('userId').first().reset_index()

# Display the first few rows of the top_recommendations DataFrame
print("\nTop Movie Recommendations (Baseline Model):")
print("===========================================")
print(top_recommendations[['userId', 'movieId', 'title', 'avg_rating']])

Average Ratings per Movie:
   movieId  avg_rating
0        1    3.920930
1        2    3.431818
2        3    3.259615
3        4    2.357143
4        5    3.071429

Top Movie Recommendations (Baseline Model):
     userId  movieId                                              title  \
0         1     1927              All Quiet on the Western Front (1930)   
1         2   131724  The Jinx: The Life and Deaths of Robert Durst ...   
2         3     5746                    Galaxy of Terror (Quest) (1981)   
3         4     3851                     I'm the One That I Want (2000)   
4         5      318                   Shawshank Redemption, The (1994)   
..      ...      ...                                                ...   
605     606    26810                               Bad Boy Bubby (1993)   
606     607      318                   Shawshank Redemption, The (1994)   
607     608     5901                                      Empire (2002)   
608     609      318                   S

**Average Ratings per Movie:**

The results displays the average ratings for the top 5 movies in the dataset.
For example, "All Quiet on the Western Front (1930)" (movieId 1) has an average rating of approximately 3.92, indicating it's well-rated on average.
**Top Movie Recommendations (Baseline Model):**

The results shows the top movie recommendations for various users based on the baseline model.
For instance, `User 2` is recommended `The Jinx: The Life and Deaths of Robert Durst` with a perfect rating of 5.0.
`User 4` is recommended `I'm the One That I Want (2000)` with a perfect rating of 5.0.
`User 5` is recommended `Shawshank Redemption, The (1994)` with an average rating of approximately 4.43, which is a highly regarded movie.
These results demonstrate the top-rated movies and personalized recommendations for users based on the baseline model. The baseline model's recommendations are primarily based on the average ratings of movies and do not take into account user-specific preferences. Further, users with higher-rated movies in the dataset tend to receive recommendations with higher ratings.

### User-Item Matrix

Let's create a user-item matrix where rows represent users, columns represent movies, and the values are the ratings. The matrix will be the foundation for collaborative filtering.